src/plugins/data/public/data_sources/datasource_selector/datasource_selectable.tsx

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import React, { useEffect } from 'react';
+import { i18n } from '@osd/i18n';
+import { EuiComboBox } from '@elastic/eui';
+import { DataSourceType, GenericDataSource } from '../datasource_services';
+import { DataSourceGroup, DataSourceSelectableProps } from './types';
+import { ISourceDataSet, IndexPatternOption } from '../datasource';
+
+type DataSourceTypeKey = 'DEFAULT_INDEX_PATTERNS' | 's3glue' | 'spark';
+
+// Mapping between datasource type and its display name.
+const DATASOURCE_TYPE_DISPLAY_NAME_MAP: Record<DataSourceTypeKey, string> = {
+  DEFAULT_INDEX_PATTERNS: 'Index patterns',
+  s3glue: 'Amazon S3',
+  spark: 'Spark',
+};
+
+type DataSetType = string | IndexPatternOption;
+
+interface DataSetWithSource {
+  ds: GenericDataSource;
+  data_sets: string[] | IndexPatternOption[];
+}
+
+export const DataSourceSelectable = ({
+  dataSources, // list of all available datasource connections.
+  dataSourceOptionList, // combo box renderable option list derived from dataSources
+  selectedSources, // current selected datasource in the form of [{ label: xxx, value: xxx }]
+  onDataSourceSelect,
+  setDataSourceOptionList,
+  onFetchDataSetError, //   onFetchDataSetError, Callback for handling dataset fetch errors. Ensure it's memoized.
+  singleSelection = true,
+}: DataSourceSelectableProps) => {
+  // This effect fetches datasets and prepares the datasource list for UI rendering.
+  useEffect(() => {
```

Comment:
None of the local functions inside `useEffect` need to be inside there to be redefined continuously. We should move them all out.

Reply:
i don't think we can use useEffectEvent hook, declaring funcs in the useEffect so that the non-reactive logic doesn't become a reactive dependency

Reply:
If we're defining reactive functions in closure (I'm not squinting at this closely), then useCallback().

Reply:
defining functions in useeffect (when needed) shouldn't be an issue, i didn't understand how is useCallback relevant? the recommended alternative would be https://react.dev/learn/separating-events-from-effects#declaring-an-effect-event